docs/rfcs/rfc-035-authorization.md

```diff
@@ -0,0 +1,759 @@
+
+### Filter
+
+For Cypher queries that begin with a `MATCH`, these rules are inserted into the `WHERE` clause directly proceeding. They are combined with filters provided in GraphQL queries using a logical `AND`.
+
+In the context of Subscriptions, these filters are applied when processing a new event, and they are also combined with user provided queries using an `AND`.
+
+Using an example rule from the example above:
+
+```gql
+type User
+  @authorization(
+    filter: [{ where: { node: { id: "$jwt.sub" } } }]
+  ) {
+  id: String!
+  name: String!
+}
+```
+
+Given the following GraphQL query:
+
+```gql
+{
+  users(where: { name: "Bob" }) {
+    id
+    name
+  }
+}
+```
+
+This will generate Cypher and parameters along the lines of the following:
+
+```cypher
+MATCH (this:User)
+WHERE this.id = $jwt.sub
+AND this.name = $name
+RETURN this { .id, .name }
+```
+
+```json
+{
+  "jwt": {
+    "sub": "123456"
+  },
+  "name": "Bob"
+}
+```
+
+### Validate
+
+Validate rules apply filters also, but throw an error in the case of reaching any data that breaches the predicate. The rules are applied using `apoc.util.validatePredicate`.
+
+By default, a validate rule is applied both before and after every operation. For example, if validating that a `User` node matches the provided JWT by checking the ID matches the subject, the odds are that you are going to want to check this is still the case following the operation.
+
+So given the following example:
+
+```gql
+type User
+  @authorization(
+    validate: [{ where: { node: { id: "$jwt.sub" } } }]
+  ) {
+  id: String!
+  name: String!
+}
+```
+
+And the following GraphQL query:
+
+```gql
+mutation {
+  updateUsers(where: { name: "Bob" }, update: { id: "654321" }) {
+    users {
+      id
+      name
+    }
+  }
+}
+```
+
+This will generate Cypher and parameters along the lines of the following:
+
+```cypher
+MATCH (this:User)
+WHERE this.name = $name
+AND apoc.util.validatePredicate(NOT (this.id = $jwt.sub), "Unauthorized", [])
+SET this.id = $update.id
+WITH this
+WHERE apoc.util.validatePredicate(NOT (this.id = $jwt.sub), "Unauthorized", [])
+RETURN this { .id, .name }
+```
+
+```json
+{
+  "jwt": {
+    "sub": "123456"
+  },
+  "name": "Bob",
+  "update": {
+    "id": "654321"
+  }
+}
+```
+
+If, following the Mutation operation, the rule is not satisfied, an error "Unauthorized" will be thrown.
+
+#### Before operations
+
+If a rule is only desired to run before operations, you will need to set `after` to an empty array:
+
+```gql
+type User
+  @authorization(
+    validate: [{ after: [], where: { node: { id: "$jwt.sub" } } }]
+  ) {
+  id: String!
+  name: String!
+}
+```
+
+Given the following GraphQL query:
+
+```gql
+{
+  users(where: { name: "Bob" }) {
+    id
+    name
+  }
+}
+```
+
+This will generate Cypher and parameters along the lines of the following:
+
+```cypher
+MATCH (this:User)
+WHERE this.name = $name
```

Comment:
I believe this syntax is dangerous, you will be able to obtain all the DB information by playing with the API.
In this example, `Unauthorized` is raised only if `this.name` is different from "Bob". This means that I can use the "Unauthorized" error as a source of truth.
If I want to know if Bob is a friend of Darrell I can write:
```graphql
{
  users(where: { name: "Bob",  friends_SOME: {
        name: "Darrell"
    }  }) {
    id
    name
  }
}
```
 And again as I know that Bob exists if not I should have received an error before, I can continue leaking information.

Reply:
I'm not sure I follow... This is taking the user's filtering arguments, and once they have been applied, we perform the authorization check. This is exactly as it performs at the moment, literally identical to `allow`.

Reply:
I think the idea is that you have a jwt with sub "A". You try to get user Bob:
`MATCH (u:User) WHERE u.name="Bob" AND ..apoc.validate.. `
Your jwt sub is checked against Bob node's ID and they will obviously not match. Therefore, validation error is thrown, indicating that Bob does indeed exist (bc otherwise apoc would not have thrown an error). 
Therefore whatever your condition is before `apoc.validate`, you will know it's true if you get an error. I think @MacondoExpress is trying to point out that if you work it out step by step you can infer a lot of information.

Reply:
However, if I remember correctly we have already discussed this is part of existing functionality so we should support it, but **strongly** suggest to use the `filter` argument instead to avoid this kind of data leaking. I might be wrong however so please correct me.

Reply:
That's exactly it, we're only doing as the user has instructed, and why `filter` is going to become such a pivotal rule!! It all comes down to good documentation communicating best practices.

Reply:
The problem is that the `authorization check` happens only if the user's where predicate returns True if it returns False then I receive an empty response. This difference in the response enables a malicious user to leak information.
Let's make an example:
```graphql
type User @authorization(
	validate: [{ after: [], where: { node: { id: "$jwt.sub" } } }]
) {
	id: String!
	name: String!
        verySecretNumber: Int
}
```
Let's consider that you Darrell have an account on this social platform with an id = 20 and your very secret number is 105.
If I with the id = 35 try to query the `users` with my id, as follow: 
```graphql
{
    users(where: { id:  20 }) {
	id
	name
        verySecretNumber
   }
}
```
I receive an error "Unauthorized" coming from the `apoc.util.validatePredicate(NOT (this.id = $jwt.sub), "Unauthorized", [])`.
But then I notice that if I make this query
```graphql
{
    users(where: { AND: [{ id:  20 }, NOT: { id: 20 }] }) {
	id
	name
        verySecretNumber
   }
}
```
I just receive an empty response.
Now I have the plan to leak your very secret number.
What I know is that if this query:
```
{
    users(where: { { id:  20, verySecretNumber: 105 } }) {
	id
	name
        verySecretNumber
   }
}
```
Returns an "Unauthorized" error then it means that your secret number is 105, if instead returns an empty response it means that I have to keep searching for it. Of course, I don't know initially your secret number but using a binary search I can find it in a couple of iterations.

Reply:
That sounds good to me it can be solvable by good documentation indeed!

Reply:
Yeah this is something I raised a few times in this `@auth` redesign. It's because we only perform the validation on data from the database. For now it should be fine as long as we document this behaviour well!

At some point in the future, I think we should consider some mechanism to throw these requests out even earlier (e.g. the user is called "Bob" but hasn't included a `where: { name: "Bob" }` filter, then throw a forbidden error before ever hitting the db).

However, this is probably out of scope for this work but it might be worth capturing in an out of scope section

Reply:
I think what I don't understand about the above proposal; if we're going to force users to use a particular filter, why would we not just apply it for them using the `filter` argument? This seems to be anecdotally the same logic as HMRC and self-assessment - "we _know_ what tax you owe, but we're going to make you calculate it yourself or we will punish you"?